my-megaparsec-project/src/Optimizer.hs

```diff
@@ -0,0 +1,192 @@
+        _ -> Prod a' b'
+  
+  Div a b -> 
+    let a' = algebraicOpt a
+        b' = algebraicOpt b
+    in case (a', b') of
+        (IntLit 0, _) -> IntLit 0         -- 0 / b = 0 (assuming b != 0)
+        (_, IntLit 1) -> a'               -- a / 1 = a
+        _ -> Div a' b'
+  
+  Neg a -> 
+    let a' = algebraicOpt a
+    in case a' of
+        Neg b -> b                        -- -(-a) = a
+        IntLit 0 -> IntLit 0              -- -0 = 0
+        _ -> Neg a'
+
+-- | Collects all variables used in an expression
+getVarsInExpr :: Expr -> S.Set Text
+getVarsInExpr (IntLit _) = S.empty
+getVarsInExpr (VarLit v) = S.singleton v
+getVarsInExpr (Sum a b) = S.union (getVarsInExpr a) (getVarsInExpr b)
+getVarsInExpr (Sub a b) = S.union (getVarsInExpr a) (getVarsInExpr b)
+getVarsInExpr (Prod a b) = S.union (getVarsInExpr a) (getVarsInExpr b)
+getVarsInExpr (Div a b) = S.union (getVarsInExpr a) (getVarsInExpr b)
+getVarsInExpr (Neg a) = getVarsInExpr a
+
+-- | Checks if an expression can be evaluated with the given substitution
+canEvalExpr :: Expr -> Subst -> Bool
+canEvalExpr expr subst = 
+  all (`M.member` subst) (S.toList $ getVarsInExpr expr)
+
+-- | Collects all variables that are modified (assigned to) in a statement
+modifiedVars :: Stmt -> S.Set Text
+modifiedVars (Assign var _) = S.singleton var
+modifiedVars (Seq s1 s2) = S.union (modifiedVars s1) (modifiedVars s2)
+modifiedVars (While _ body) = modifiedVars body
+modifiedVars (Write _) = S.empty
+modifiedVars (Read var) = S.singleton var
+
+
+-- | Computes the resulting substitution after evaluating a statement
+evalStaticStmt :: Stmt -> Subst -> Subst
+evalStaticStmt (Assign var expr) subst = 
+  if canEvalExpr expr subst
+  then M.insert var (evalExpr expr subst) subst
+  else M.delete var subst
+evalStaticStmt (Seq s1 s2) subst =
+  let subst' = evalStaticStmt s1 subst
+  in evalStaticStmt s2 subst'
+-- For a while loop, any variable modified in the body becomes unknown
+evalStaticStmt (While _ body) subst = foldr M.delete subst (S.toList $ modifiedVars body)
+evalStaticStmt (Write _) subst = subst
+evalStaticStmt (Read var) subst = M.delete var subst
+
+-- | Checks if a statement can be evaluated with the given substitution
+canEvalStmt :: Stmt -> Subst -> Bool
+canEvalStmt (Assign _ expr) subst = canEvalExpr expr subst
+canEvalStmt (Seq s1 s2) subst =
+  let canEval1 = canEvalStmt s1 subst
+      -- If we can evaluate s1, use its result for s2's evaluation
+      subst' = if canEval1 
+               then evalStaticStmt s1 subst
+               else foldr M.delete subst (S.toList $ modifiedVars s1)
+  in canEval1 && canEvalStmt s2 subst'
+canEvalStmt (While cond _) subst = False -- to complicated for now
+canEvalStmt (Write expr) subst = canEvalExpr expr subst
+canEvalStmt (Read _) _ = False
+
+-- | Optimizes a statement by evaluating constant expressions
+-- | Returns the optimized statement and the updated substitution map
+optStmt :: Stmt -> Subst -> (Stmt, Subst)
```

Comment:
You can make this function to be a monadic computation which tracks changes in the substitution